internal/etag/reader.go

```diff
@@ -153,3 +167,41 @@ type VerifyError struct {
 func (v VerifyError) Error() string {
 	return fmt.Sprintf("etag: expected ETag %q does not match computed ETag %q", v.Expected, v.Computed)
 }
+
+// UUIDHash - use uuid to make md5sum
+type UUIDHash struct {
+	uuid string
+}
+
+// Write -  implement hash.Hash Write
+func (u UUIDHash) Write(p []byte) (n int, err error) {
+	return len(p), nil
+}
+
+// Sum -  implement md5.Sum
+func (u UUIDHash) Sum(b []byte) []byte {
+	hasher := md5.New()
+	return hasher.Sum([]byte(u.uuid))
```

Comment:
`u.uuid` is the string encoding, not the UUID values. Here is how you get the raw UUID:

```Go
// mustGetUUIDBytes - get a random UUID as 16 bytes unencoded.
func mustGetUUIDBytes() []byte {
	u, err := uuid.NewRandom()
	if err != nil {
		logger.CriticalIf(GlobalContext, err)
	}
	return u[:]
}
```

Alternatively you can parse the UUID, but that seems wasteful.

Reply:
Yes, It comes from the ctx. `u.uuid = mustGetUUID()`. This is external random generation, I think it can be decided externally here. If the UUID passed in is the same, the value of ETAG will be the same. @klauspost

Reply:
Yes. That is the *encoded* string, not a 16 byte.

The encoded version looks like this: `c59a75c1-e28d-4d07-a3ec-1e7ff63b8ceb` - the UUID values are just 16 values in a `[]byte`  - as is co-incidentally the MD5 before encoding.